combinetf2/fitter.py

```diff
@@ -539,7 +539,136 @@ def global_impacts_parms(self):
+                            profile=True, compute_norm=False, full=False
+                        )
+                        lbeta, _ = self._compute_lbeta(beta)
+                        val = lc + lbeta
+                    else:
+                        val = lc
+            pdldbeta = t1.gradient(val, self.ubeta)
+        if self.externalCovariance:
+            pd2ldbeta2_matrix = t2.jacobian(pdldbeta, self.ubeta)
+            pd2ldbeta2 = tf.linalg.LinearOperatorFullMatrix(
+                pd2ldbeta2_matrix, is_self_adjoint=True
+            )
+        else:
+            # pd2ldbeta2 is diagonal, so we can use gradient instead of jacobian
+            pd2ldbeta2_diag = t2.gradient(pdldbeta, self.ubeta)
+            pd2ldbeta2 = tf.linalg.LinearOperatorDiag(
+                pd2ldbeta2_diag, is_self_adjoint=True
+            )
+        return pd2ldbeta2
+
+    def _chi2(
+        self,
+        fun,
+        ndf_reduction=0,
+        profile=False,
+    ):
+
+        def compute_derivatives(dvars):
+            with tf.GradientTape(watch_accessed_variables=False) as t:
+                t.watch(dvars)
+                expected = self._compute_expected(
+                    fun,
+                    inclusive=True,
+                    profile=profile,
+                    full=False,
+                    need_observables=True,
+                )
+                observed = fun(None, self.nobs)
+                residuals = expected - observed
+
+                residuals_flat = tf.reshape(residuals, (-1,))
+            jacs = t.jacobian(
+                residuals_flat,
+                dvars,
+            )
+            return residuals, *jacs
+
+        if self.binByBinStat:
+            dvars = [self.x, self.ubeta, self.nobs]
+            residuals, pdresdx, pdresdbeta, pdresdobs = compute_derivatives(dvars)
+        else:
+            dvars = [self.x, self.nobs]
+            residuals, pdresdx, pdresdobs = compute_derivatives(dvars)
+            pdresdbeta = None
+
+        cov_dresdx = tf.matmul(self.cov, pdresdx, transpose_b=True)
+
+        res_cov = pdresdx @ cov_dresdx
+
+        if pdresdbeta is not None:
+            pd2ldbeta2 = self._pd2ldbeta2(profile)
+            pd2ldbeta2_pdresdbeta = pd2ldbeta2.solve(pdresdbeta, adjoint_arg=True)
+            res_cov += pdresdbeta @ pd2ldbeta2_pdresdbeta
+
+        # data stat part
+        if profile:
+            with tf.GradientTape() as t2:
+                t2.watch(self.nobs)
+                with tf.GradientTape() as t1:
+                    t1.watch(self.nobs)
+                    val = self._compute_loss()
+                grad = t1.gradient(val, self.x, unconnected_gradients="zero")
+            pd2ldxdnobs = t2.jacobian(grad, self.nobs)
+
+            dxdnobs = -self.cov @ pd2ldxdnobs
+
+            dresdnobs = pdresdobs + pdresdx @ dxdnobs
+
+            if self.externalCovariance:
+                data_cov = tf.math.sqrt(tf.linalg.inv(self.data_cov_inv))
```

Comment:
this isn't correct since this is an element-wise sqrt.  You would need the cholesky decomposition, but actually this is a waste of calculation because you could just directly calculate

```python
res_cov += dresdnobs @ data_cov @ dresdnobs.T
```
(or the appropriate variation with solve etc)

For consistency can do the same for the case for the non-externalCovariance case (e.g. using tf.linalg.LinearOperatorDiag to keep the logic the same while avoiding unnecessary calculations)